cdap-watchdog/src/main/java/io/cdap/cdap/metrics/process/MessagingMetricsProcessorManagerService.java

```diff
@@ -0,0 +1,106 @@
+                                          MetricStore metricStore,
+                                          @Assisted Set<Integer> topicNumbers,
+                                          @Assisted MetricsContext metricsContext,
+                                          @Assisted Integer instanceId) {
+    this(cConf, metricDatasetFactory, messagingService,
+         schemaGenerator, readerFactory, metricStore, topicNumbers, metricsContext,
+         TimeUnit.SECONDS.toMillis(cConf.getInt(Constants.Metrics.METRICS_MINIMUM_RESOLUTION_SECONDS)), instanceId);
+  }
+
+  @VisibleForTesting
+  MessagingMetricsProcessorManagerService(CConfiguration cConf,
+                                          MetricDatasetFactory metricDatasetFactory,
+                                          MessagingService messagingService,
+                                          SchemaGenerator schemaGenerator,
+                                          DatumReaderFactory readerFactory,
+                                          MetricStore metricStore,
+                                          Set<Integer> topicNumbers,
+                                          MetricsContext metricsContext,
+                                          long metricsProcessIntervalMillis,
+                                          int instanceId) {
+    this.metricsWriters = new ArrayList<>();
+    this.metricsProcessorServices = new ArrayList<>();
+    this.metricsWriters.add(new MetricStoreMetricsWriter(metricStore, metricsContext));
+
+    for (MetricsWriter metricsExtension : this.metricsWriters) {
+      metricsProcessorServices.add(new MessagingMetricsProcessorService(cConf,
+                                                                        metricDatasetFactory,
+                                                                        messagingService,
+                                                                        schemaGenerator,
+                                                                        readerFactory,
+                                                                        metricsExtension,
+                                                                        topicNumbers,
+                                                                        metricsContext,
+                                                                        metricsProcessIntervalMillis,
+                                                                        instanceId));
+
+    }
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    for (MessagingMetricsProcessorService processorService : metricsProcessorServices) {
+      if (!processorService.isRunning()) {
+        processorService.startAndWait();
+      }
+    }
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    for (MessagingMetricsProcessorService processorService : metricsProcessorServices) {
+      processorService.stopAndWait();
```

Comment:
Catch exception raised by the `stopAndWait()` and throw it/them (use Throwable.addSuppressed to aggregate them) after the for loop.

Reply:
Done